src/main/java/org/noureddine/joularjx/monitor/MonitoringHandler.java

```diff
@@ -247,47 +258,91 @@ private Map<Thread, Map<CallTree, Integer>> extractCallTreesStats(Map<Thread, Li
-        long totalThreadsCpuTime = 0;
-        for (var entry : methodsStats.entrySet()) {
-            long threadCpuTime = threadBean.getThreadCpuTime(entry.getKey().getId());
-
-            threadCpuTime *= entry.getValue().values().stream().mapToDouble(i -> i).sum() / sampleIterations;
-
-            // If thread already monitored, then calculate CPU time since last time
-            threadCpuTime = threadsCpuTime.merge(entry.getKey().getId(), threadCpuTime,
-                    (present, newValue) -> newValue - present);
-
-            totalThreadsCpuTime += threadCpuTime;
-        }
-        return totalThreadsCpuTime;
-    }
-
-    /**
-     * Returns for each thread (PID) it's percentage of CPU time used
-     * @param threadsCpuTime a map of CPU time per PID
-     * @param totalThreadsCpuTime the CPU time used by all the monitored threads
-     * @param processEnergy the energy consumed by the process
-     * @return for each PID, the percentage of CPU time used by the associated thread
+     * 
+     * @param methodsStats a map of method occurrences for each thread
+     * @param threadsCpuTime a map of CPU time per PID, contains the cpu time for each tread, resulting from the last call to getThreadCpuTime(threadId)
+     * @param processEnergy the energy consumed by the process in the last monitoring period
+     * @return for each PID, the percentage of energy used by the associated thread
      */
-    private Map<Long, Double> getThreadsCpuTimePercentage(Map<Long, Long> threadsCpuTime,
-                                                          long totalThreadsCpuTime,
-                                                          double processEnergy) {
-        Map<Long, Double> threadsPower = new HashMap<>();
-        for (var entry : threadsCpuTime.entrySet()) {
-            double percentageCpuTime = (entry.getValue() * 100.0) / totalThreadsCpuTime;
-            double threadPower = processEnergy * (percentageCpuTime / 100.0);
-            threadsPower.put(entry.getKey(), threadPower);
-        }
-        return threadsPower;
-    }
-
+    private Map<Long, Double> getThreadsCpuTimePercentage(Map<Thread, Map<String, Integer>> methodsStats,
+			Map<Long, Long> threadsCpuTime, double processEnergy) {
+		Map<Long, Double> threadsCpuTimePercentage = new HashMap<Long, Double>();
+    	
+		Map<Long, Double> actualThreadsCpuTime = new HashMap<Long, Double>(); 
+		double totalThreadsCpuTime = 0;
+		// first compute the proportion of cpu time for each thread in the last sampling period
+		for (Entry<Thread, Map<String, Integer>> threadEntry : methodsStats.entrySet()) {
+			long threadId = threadEntry.getKey().getId();
+			long currentThreadCpuTime = threadBean.getThreadCpuTime(threadId);
+			long previousThreadCpuTime = threadsCpuTime.getOrDefault(threadId, 0l);
+			if (currentThreadCpuTime < 0) { // thread has quit
+				// TODO ignore last sampling period??

```

Comment:
The sampling rate is by default 10ms and it is defined in the config.properties file (stack-monitoring-sample-rate), and it's used in the ```sampleRateMilliseconds``` variable.
I'd say, for now, we can ignore when thread has quit.
So your proposition seems fine, although I think 1 nanosecond might be too low in most heavy threads. Maybe use 1 millisecond or a fraction of the sampling rate, or more cleanly, ```this.sampleRateMilliseconds / 10```